library/src/main/java/com/meisolsson/githubsdk/core/GitHubPaginationInterceptor.java

```diff
@@ -34,37 +39,50 @@ public Response intercept(Chain chain) throws IOException {
         if (response.isSuccessful()) {
             String linkHeader = response.header("link");
             final boolean isArray = response.peekBody(1).string().equals("[");
-            final ArrayList<String> pageJson;
+            String pageJson = null;
 
             if (linkHeader != null) {
-                pageJson = new ArrayList<>();
+                ArrayList<String> pageItems = new ArrayList<>();
                 String[] links = linkHeader.split(",");
                 for (String link : links) {
                     String[] pageLink = link.split(";");
                     String page = Uri.parse(pageLink[0].replaceAll("[<>]", "")).getQueryParameter("page");
                     String rel = pageLink[1].replaceAll("\"", "").replace("rel=", "");
 
-                    if (page != null)
-                        pageJson.add(String.format("\"%s\":\"%s\"", rel.trim(), page));
+                    if (page != null) {
+                        pageItems.add(String.format("\"%s\":\"%s\"", rel.trim(), page));
+                    }
+                }
+                if (!pageItems.isEmpty()) {
+                    pageJson = TextUtils.join(",", pageItems);
                 }
-            } else {
-                pageJson = null;
             }
 
-            final boolean hasPageInfo = pageJson != null && !pageJson.isEmpty();
-            if (isArray || hasPageInfo) {
-                final String json;
-                if (isArray) {
-                    json = String.format("{%s%s\"items\":%s}",
-                            hasPageInfo ? TextUtils.join(",", pageJson) : "",
-                            hasPageInfo ? ", " : "",
-                            response.body().string());
-                } else {
-                    json = String.format("{%s, %s",
-                            TextUtils.join(", ", pageJson),
-                            response.body().string().substring(1));
+            if (isArray || pageJson != null) {
+                MediaType contentType = response.body().contentType();
+                Charset charset = contentType != null ? contentType.charset(Util.UTF_8) : Util.UTF_8;
+                BufferedSource source = response.body().source();
+                Buffer newResponse = new Buffer();
+
+                try {
+                    source.request(Long.MAX_VALUE);
+
+                    if (isArray) {
+                        newResponse.writeString("{\"items\":", charset);
+                        newResponse.write(source.buffer(), source.buffer().size());
```

Comment:
Why not use `Buffer.writeAll(Source source)`guessing that would allow us to remove `source.request(...)`

Reply:
Forget that, saw why that's used further down.